### PR TITLE
[ARM-ARCH] Ensure that any byrefs created in genCodeForIndexAddr cannot point outside of the array

### PR DESCRIPTION
**Pull request by briansull**
Ensure that any byrefs created in genCodeForIndexAddr cannot point outside of the array
Used the scaled add instructions for more powers of two
Only use the safe byref sequence when generating fully interruptible code.

### CODE REVIEW

**Comment by briansull** [comment]
Fixes #17823

**Comment by briansull** [comment]
@dotnet-bot Test OSX10.12 x64 Checked Innerloop Build and Test

**Comment by briansull** [comment]
@BruceForstall  PTAL
@sdmaclea  PTAL

**Comment by briansull** [comment]
Currently we are fairly cautious when folding constants into array LEA instructions., so I couldn't identify an error with our current codegen. That could change in the future, so we should fix this now to prevent any problems in the future when we do more address folding operations.

File: src/jit/codegenarmarch.cpp [hunk 1]
```diff
@@ -3500,11 +3486,20 @@ void CodeGen::genCodeForStoreBlk(GenTreeBlk* blkOp)
 void CodeGen::genScaledAdd(emitAttr attr, regNumber targetReg, regNumber baseReg, regNumber indexReg, int scale)
 {
     emitter* emit = getEmitter();
+    if (scale == 0)
+    {
+        // target = base + index
+        getEmitter()->emitIns_R_R_R(INS_add, attr, targetReg, baseReg, indexReg);
+    }
+    else
+    {
```
Comment (sdmaclea) [inline src/jit/codegenarmarch.cpp]: Why is it worth special casing scale == 0?

File: src/jit/codegenarmarch.cpp [hunk 2]
```diff
@@ -1711,38 +1711,24 @@ void CodeGen::genCodeForIndexAddr(GenTreeIndexAddr* node)
         genJumpToThrowHlpBlk(genJumpKindForOper(GT_GE, CK_UNSIGNED), SCK_RNGCHK_FAIL, node->gtIndRngFailBB);
     }
 
-    // Compute the address of the array element.
-    switch (node->gtElemSize)
+    // Can we use a ScaledAdd instruction?
+    //
+    if (isPow2(node->gtElemSize) && (node->gtElemSize <= 32768))
```
Comment (sdmaclea) [inline src/jit/codegenarmarch.cpp]: Why the magic number 32768?  Looks like arm64 supports shifts of 0-31/63.  Arm32 at first glance should be similar.

File: src/jit/codegenarmarch.cpp [hunk 3]
```diff
@@ -3500,11 +3486,20 @@ void CodeGen::genCodeForStoreBlk(GenTreeBlk* blkOp)
 void CodeGen::genScaledAdd(emitAttr attr, regNumber targetReg, regNumber baseReg, regNumber indexReg, int scale)
```
Comment (sdmaclea) [inline src/jit/codegenarmarch.cpp]: This function name and the variable name scale are unfortunate.  Scale implies multiply not shift.

File: src/jit/codegenarmarch.cpp (hunk 2 above)
Comment (briansull) [reply src/jit/codegenarmarch.cpp]: I could increase this to a larger power of two, but would still need some upper bound check, 
so I picked this value as a reasonable limit check.

**Comment by briansull** [comment]
Small number of difsf mostly due to array of 64-byte structs using shiftedAdd sequence insteal of multiplyAdd sequence. 
Three methods now use the longer safe sequence:
17.99:           Regression # 1 goes from   732 to   740, diff of     8 ( 1.08%) :: Microsoft.VisualBasic.Interaction:GetAllSettings(ref,ref):ref
18.02:           Regression # 1 goes from   288 to   292, diff of     4 ( 1.37%) :: System.DirectoryServices.ActiveDirectory.ActiveDirectorySchedule:get_RawSchedule():ref:this
18.06:           Regression # 1 goes from   712 to   716, diff of     4 ( 0.56%) :: System.Web.Management.WebBaseEvent:IncrementTotalCounters(int,int):this

File: src/jit/codegenarmarch.cpp (hunk 1 above)
Comment (briansull) [reply src/jit/codegenarmarch.cpp]: The emitter currently doesn't allow 0 to be used with INS_OPTS_LSL:

```
        else if (insOptsAluShift(opt))
        {
            // imm should be non-zero and in [1..63]
            assert(isValidImmShift(imm, size) && (imm != 0));
            fmt = IF_DR_3B;
        }
```

**Pull request merged**